src/server/frontend_wayland/foreign_toplevel_manager_v1.h

```diff
@@ -36,8 +39,30 @@ auto create_foreign_toplevel_manager_v1(
     wl_display* display,
     std::shared_ptr<shell::Shell> const& shell,
     std::shared_ptr<Executor> const& wayland_executor,
-    std::shared_ptr<SurfaceStack> const& surface_stack)
+    std::shared_ptr<SurfaceStack> const& surface_stack,
+    std::shared_ptr<DesktopFileManager> const& desktop_file_manager)
 -> std::shared_ptr<wayland::ForeignToplevelManagerV1::Global>;
+
+class GDesktopFileCache : public DesktopFileCache
+{
+public:
+    GDesktopFileCache(std::shared_ptr<MainLoop> const& main_loop);
+    ~GDesktopFileCache() override = default;
+    std::shared_ptr<DesktopFile> lookup_by_app_id(std::string const&) const override;
+    std::shared_ptr<DesktopFile> lookup_by_wm_class(std::string const&) const override;
+    std::vector<std::shared_ptr<DesktopFile>> const& get_desktop_files() const override;
+    void refresh_app_cache();
+
+private:
+    mutable std::recursive_mutex update_mutex;
```

Comment:
Hm. `recursive_mutex` is somewhat of a code smell. It's also not clear (from here) what it's protecting.

You could, perhaps, use `mir::Synchronised` (or two different `Synchronised`s; uncontended lock acqisition is *super* cheap) here, or at least move the mutex down to the fields it's protecting.

Reply:
It's protecting access to the cached files, maps, etc. Basically, the update of those variable can happen in another thread (when we are notified that the files changed), so I'm trying to avoid the situation where another thread queries for the filed and receives an half-built list.

Reply:
And recursive_mutex was used for bad reasons. We can chat about this in your morning ;)